specs/plugin-extensibility.md

```diff
@@ -0,0 +1,277 @@
+           3. Check that the `response.certificateChain` confirms to [Certificate Requirements].
+        4. Assemble the JWS Signature envelope using `response.signature`, `response.signingAlgorithm` and `response.certificateChain`. Notation may also generate and include timestamp signature in this step.
+        5. Generate a signature manifest for the given signature envelope.
+    2. Else if plugin supports capability `SIGNATURE_ENVELOPE_GENERATOR` *(covered in next section)*
+    3. Return an error
+
+***generate-signature***
+
+*Request*
+
+```json
+{
+  "contract-version" : <major-version.minor-version>,
+
+  // Complete key definition from
+  // /notation/config.json /signingKeys/keys with matching key name
+  "key" : <key definition>,
+
+  "payload" : <Base64 encoded payload to be signed>
+}
+```
+
+*key* : Required field that contains the complete key definition, which includes friendly name (`name`), key identifier (`id`) and any additional custom fields required by the plugin.
+
+*payload* : Required field that contains payload to be signed. This is currently the JWSPayload that includes the subject descriptor and other JWS claims.
+
+*Response*
+All response attributes are required.
+
+```json
+{    
+  // The same key id as passed in the request.
+  "keyId" : <key id>,
+  "signature" : <Base64 encoded signature>,
+  "signingAlgorithm" : <signing algorithm>
+  "certificateChain": ["Base64(DER(leafCert))","Base64(DER(intermediateCACert))","Base64(DER(rootCert))"]
+}
+```
+
+*signingAlgorithm* : One of following [supported signing algorithms](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#algorithm-selection), Notation uses this validate the signature, and to set the appropriate attribute in signature envelope (e.g. JWS `alg`). `RSASSA_PSS_SHA_256`, `RSASSA_PSS_SHA_384`, `RSASSA_PSS_SHA_512`,  `ECDSA_SHA_256`, `ECDSA_SHA_384`, `ECDSA_SHA_512`.
+
+*certificateChain* : Ordered list of certificates starting with leaf certificate and ending with root certificate.
+
+*Error codes*
+
+* VALIDATION_ERROR - Any of the required request fields was empty, or a value was malformed/invalid. Includes condition where the key referenced by `keyId` was not found.
+* UNSUPPORTED_CONTRACT_VERSION - The contract version used in the request is unsupported.
+* ACCESS_DENIED - Authentication/authorization error to use given key.
+* TIMEOUT - The operation to generate signature timed out and can be retried by Notation.
+* THROTTLED - The operation to generate signature was throttles and can be retried by Notation.
+* ERROR - Any general error that does not fall into previous error categories.
```

Comment:
These error codes are good to be surfaced out but IMO notation should not perform any retry action. In case of throttled requests, it's possible request can only be retried after a certain time defined in `retry-after` header. The plugin can decide to (1) retry the request internally before giving up and returning an error or (2) return the `retry-after` header info as part of error message, so the user can retry the request after the time has elapsed.

Reply:
This also makes me think there should be a context defined with timeout as part of the request to the plugin. We can use [CommandContext](https://pkg.go.dev/os/exec#CommandContext) to ensure the request doesn't hang indefinitely

Reply:
Agreed, Notation should not retry. The errors would be surfaced to caller invoking Notation, and it can decide to retry if needed.

Reply:
Can you provide more details about the context with timeout as part of the request, maybe with an example, I didn't understand that.